ng update deletes node_modules before ensuring it will succeed

# 🐞 Bug report

### Command (mark with an `x`)

`ng update @angular/cli @angular/cdk ...`


- [ ] new
- [ ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [x] update
- [ ] lint
- [ ] extract-i18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

Not sure. Might be half bug, half feature request.

### Description

When updating from angular 12 to 13, the following sequence of events happened:

1. `ng update` to determine dependencies
2. `ng update ...` 
  1. after determining dependencies and updating my package.json file, the script seems to delete node_modules
  2. the `npm install` step triggered by `ng update` fails due to a peer dependency conflict 
  3. *this is the FR/bug* this part of the script should do a dry run and abort *before* deleting `node_modules` if the `npm install` with the updated package.json file will fail. Ideally add more details and suggestions here, and roll back the change to `package.json`.
3. Now re-running `ng update` a second time will note no installed dependencies, because apparently the script checks `node_modules` and not `package.json`. 
4. The actual fix for the deeper bug is some conflict in peer dependencies between the older 12 installation of architect and build-angular and the updated `package.json` file. Reverting `package.json` to head (the 12 installation), running `npm install` to re-install the now-missing `node_modules` directory, then deleting `package-lock.json`, will allow the `ng update` command to run to completion. If this is a common issue, adding some help text in that regard would be nice.

## 🌍 Your Environment

*NOTE* this is post-upgrade, the bug happened during a 12->13 upgrade.

<pre><code>
Angular CLI: 13.0.2
Node: 16.10.0
Package Manager: npm 7.24.0
OS: linux x64

Angular: 13.0.1
... animations, cdk, common, compiler, compiler-cli, core, forms
... material, platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1300.2
@angular-devkit/build-angular   13.0.2
@angular-devkit/core            13.0.2
@angular-devkit/schematics      13.0.2
@angular/cli                    13.0.2
@angular/fire                   7.2.0
@schematics/angular             13.0.2
rxjs                            6.6.7
typescript                      4.4.4

</code></pre>

**Anything else relevant?**

Same issue with update from Angular 10 to 11. node_modules folder gone! :( :( 

From `https://update.angular.io/?l=2&v=10.0-12.0` 
`npx @angular/cli@11 update @angular/core@11 @angular/cli@11 `
Only issues a one-line error message about incomplete repository. But I am not event using git.
Stackoverflow suggests to add --allow-dirty
It runs with that. I get
`npm WARN old lockfile The package-lock.json file was created with an old version of npm,`
Not sure if that is important.
Then plenty of errors related to @angular-devkit/build-angular@0.1000.4 
```
npm ERR! While resolving: workspace@0.0.0
npm ERR! Found: @angular-devkit/build-angular@0.1000.4
npm ERR! node_modules/@angular-devkit/build-angular
npm ERR!   dev @angular-devkit/build-angular@"~0.1102.15" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! dev @angular-devkit/build-angular@"~0.1102.15" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: @angular/compiler-cli@11.2.14
npm ERR! node_modules/@angular/compiler-cli
npm ERR!   peer @angular/compiler-cli@"^11.0.0 || ^11.2.0-next" from @angular-devkit/build-angular@0.1102.15
npm ERR!   node_modules/@angular-devkit/build-angular
npm ERR!     dev @angular-devkit/build-angular@"~0.1102.15" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR! See C:\Users\andreas2\AppData\Local\npm-cache\eresolve-report.txt for a full report.
```
[eresolve-report.txt](https://github.com/angular/angular-cli/files/7641418/eresolve-report.txt)

--- a/packages/angular/cli/src/commands/update/cli.ts
+++ b/packages/angular/cli/src/commands/update/cli.ts
@@ -541,6 +541,14 @@ export default class UpdateCommandModule extends CommandModule<UpdateCommandArgs
       return 1;
     }
 
+    const packageJsonPath = path.join(this.context.root, 'package.json');
+    let originalPackageJsonContent: string | undefined;
+    try {
+      originalPackageJsonContent = await fs.readFile(packageJsonPath, 'utf8');
+    } catch {
+      // Ignore backup errors.
+    }
+
     try {
       await applyUpdatePlan(this.context.root, plan, logger);
     } catch (error) {
@@ -596,6 +604,16 @@ export default class UpdateCommandModule extends CommandModule<UpdateCommandArgs
         Module._pathCache = Object.create(null);
       }
     } catch (e) {
+      if (originalPackageJsonContent !== undefined) {
+        try {
+          await fs.writeFile(packageJsonPath, originalPackageJsonContent, 'utf8');
+          logger.info('Restored package.json to its original state.');
+        } catch (restoreError) {
+          assertIsError(restoreError);
+          logger.error(`Failed to restore package.json: ${restoreError.message}`);
+        }
+      }
+
       if (e instanceof CommandError) {
         return 1;
       }

--- a/packages/angular/cli/src/commands/update/cli.ts
+++ b/packages/angular/cli/src/commands/update/cli.ts
@@ -567,6 +567,11 @@ export default class UpdateCommandModule extends CommandModule<UpdateCommandArgs
     const tasks = new Listr([
       {
         title: 'Cleaning node modules directory',
+        skip() {
+          return packageManager.name !== 'npm'
+            ? 'Cleaning not required for this package manager.'
+            : false;
+        },
         async task(_, task) {
           try {
             await fs.rm(path.join(commandRoot, 'node_modules'), {

